segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/ThrottlerCalculator.java

```diff
@@ -0,0 +1,93 @@
+import io.pravega.common.MathHelpers;
+import io.pravega.segmentstore.storage.QueueStats;
+import java.util.function.Supplier;
+
+/**
+ * Helper class that provides methods for calculating various OperationProcessor delays, to be used for batching operations
+ * or throttling requests.
+ */
+class ThrottlerCalculator {
+    @VisibleForTesting
+    static final int MAX_BATCHING_DELAY_MILLIS = 50;
+    @VisibleForTesting
+    static final int MAX_THROTTLING_DELAY_MILLIS = 10000;
+    @VisibleForTesting
+    static final double MAX_CACHE_UTILIZATION = 1.25; // 125% of the cache capacity.
+
+    private final Supplier<QueueStats> getQueueStats;
+    private final Supplier<Double> getCacheUtilization;
+
+    /**
+     * Creates a new instance of the ThrottlerCalculator class.
+     *
+     * @param getQueueStats       A Supplier that, when invoked, returns a QueueStats object representing the most recent
+     *                            statistics about the DurableDataLog write queue.
+     * @param getCacheUtilization A Supplier that, when invoked, returns a non-negative number representing the cache
+     *                            utilization (calculated as the ratio of used cache to total cache available). May be
+     *                            greater than 1 if too much cache is used (and thus throttling is required).
+     */
+    ThrottlerCalculator(Supplier<QueueStats> getQueueStats, Supplier<Double> getCacheUtilization) {
+        this.getQueueStats = Preconditions.checkNotNull(getQueueStats, "getQueueStats");
+        this.getCacheUtilization = Preconditions.checkNotNull(getCacheUtilization, "getCacheUtilization");
+    }
+
+    /**
+     * Determines whether there is a need to introduce a throttling delay. If this method returns false, then
+     * getThrottlingDelayMillis() will return 0.
+     *
+     * @return True if throttling is required, false otherwise.
+     */
+    boolean isThrottlingRequired() {
+        return this.getCacheUtilization.get() > 1;
+    }
+
+    /**
+     * Calculates the amount of time to wait before processing more operations from the queue in order to relieve pressure
+     * on downstream components. This is based on ReadIndex statistics, mainly cache utilization ratio.
+     *
+     * @return The amount of time, in milliseconds. If isThrottlingRequired() returns false, this will return 0, otherwise
+     * it will return a positive value.
+     */
+    int getThrottlingDelayMillis() {
+        double cacheUtilization = this.getCacheUtilization.get();
+        if (cacheUtilization <= 1) {
+            // Cache utilization is below 100%. No need for throttling.
+            return 0;
+        } else {
+            // We are exceeding the cache capacity. Start throttling proportionally to the amount of excess, but make sure
+            // we don't exceed the max allowed.
+            return (int) (Math.min(1.0, (cacheUtilization - 1) / (MAX_CACHE_UTILIZATION - 1)) * MAX_THROTTLING_DELAY_MILLIS);
```

Comment:
Linear throttling seems like a good idea. But I not so sure about these constants.
If we had a cache capacity of say 1GB that would mean the calculation would mean that:
At 250mb over it would throttle for 10000ms, at 25mb over it would throttle for 1000ms, 2.5mb over it would throttle for 100ms, 250kb over it would throttle at 10ms and at 25kb over it would throttle at 1ms.

This seems a bit sudden on the low end. Particularly as it does not throttle at all before reaching capacity. So for instance it might not throttle at all and then go 1MB over from a single BK operation and suddenly be throttling at 40ms. 

What if we changed MAX_CACHE_UTILIZATION to be 2.0.? Then 1MB would be 10ms, 10mb would be 100ms, 100mb would be 1s, etc.

Reply:
I see your point. But let's see how this would work in practice. Our Cache has a physical hard limit (whether it be main memory or some dedicated SSD (assuming we use Rocks - which we do)). We will configure our "max cache size" (via the config file) to something around 80% of that, which means we can use it freely up to that limit. When we reach that, we start evicting evictable data, at which point we still work fine. But once we cannot evict any more data, we will begin throttling, at the same time allowing for an excess to this cache (so it will go beyond 80% of physical capacity, up to 100%). When it reaches 100% the system will likely crash as all cache insert operations will fail (let it be OOM Exceptions or some other fatal Rocks exceptions due to disks full). That's the reasoning behind such a steep curve.

Should we do something softer, as you proposed, then (with MAX_CACHE_UTILIZATION == 2.0), we will be setting our cache limit to 50% of the max physical capacity). This way we do indeed have a bigger buffer until we reach a critical point, but we will be essentially wasting 50% of our cache capacity in the vast majority of time (we only expect to throttle in exceptional cases, when Tier2 really does have a problem). 

IMO, I feel that setting cache limit to 80% of physical capacity with a 25% excess alllowance is a better tuning for the system rather than 50% with 100% excess.